File: packages/oxlint-plugin-react-doctor/src/plugin/rules/performance/no-large-animated-blur.ts
import {
  BLUR_VALUE_PATTERN,
  LARGE_BLUR_THRESHOLD_PX,
  MOTION_ANIMATE_PROPS,
} from "../../constants/style.js";
import { defineRule } from "../../utils/define-rule.js";
import type { Rule } from "../../utils/rule.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

export const noLargeAnimatedBlur = defineRule<Rule>({
  id: "no-large-animated-blur",
  title: "Large animated blur",
  tags: ["test-noise"],
  severity: "warn",
  recommendation:
    "Keep the blur under 10px, or blur a smaller element. Big blurs use a lot more GPU memory as the element grows",
  create: (context: RuleContext) => ({
    JSXAttribute(node: EsTreeNodeOfType<"JSXAttribute">) {
      if (!isNodeOfType(node.name, "JSXIdentifier")) return;
      if (node.name.name !== "style" && !MOTION_ANIMATE_PROPS.has(node.name.name)) return;
      if (!isNodeOfType(node.value, "JSXExpressionContainer")) return;

      const expression = node.value.expression;
      if (!isNodeOfType(expression, "ObjectExpression")) return;

      for (const property of expression.properties ?? []) {
        if (!isNodeOfType(property, "Property")) continue;
        const key = isNodeOfType(property.key, "Identifier") ? property.key.name : null;
        if (key !== "filter" && key !== "backdropFilter" && key !== "WebkitBackdropFilter")
          continue;
        if (!isNodeOfType(property.value, "Literal") || typeof property.value.value !== "string")
          continue;

        const match = BLUR_VALUE_PATTERN.exec(property.value.value);
        if (!match) continue;

        const blurRadius = Number.parseFloat(match[1]);
        if (blurRadius > LARGE_BLUR_THRESHOLD_PX) {
          context.report({
            node: property,
            message: `Large animated blurs can use significant GPU memory on phones because blur(${blurRadius}px) gets heavier as the blur and element grow. Use a smaller blur or a smaller element.`,
          });
        }
      }
    },
  }),
});

File: native/oxlint/rules/no-large-animated-blur.rs
use std::borrow::Cow;

use lazy_regex::{Lazy, Regex, lazy_regex};
use oxc_ast::ast::{
    JSXAttributeName, JSXAttributeValue, ObjectExpression, ObjectProperty, ObjectPropertyKind,
    PropertyKey,
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{
    AstNode,
    context::{ContextHost, LintContext},
    rule::Rule,
};

const MOTION_ANIMATE_PROPS: [&str; 8] = [
    "animate",
    "initial",
    "exit",
    "whileHover",
    "whileTap",
    "whileFocus",
    "whileDrag",
    "whileInView",
];
static BLUR_VALUE_PATTERN: Lazy<Regex> = lazy_regex!(r"blur\((\d+(?:\.\d+)?)px\)");

#[derive(Debug, Default, Clone)]
pub struct NoLargeAnimatedBlur;

declare_oxc_lint!(
    /// Disallow large animated CSS blurs.
    NoLargeAnimatedBlur,
    react_doctor_native,
    perf,
    version = "0.1.0",
    short_description = "Disallow large animated CSS blurs.",
);

impl Rule for NoLargeAnimatedBlur {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_test_noise_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        match node.kind() {
            AstKind::JSXAttribute(attribute) => check_motion_attribute(attribute, ctx),
            AstKind::CallExpression(call_expression) => {
                check_web_animation_call(call_expression, ctx);
            }
            _ => {}
        }
    }
}

fn check_motion_attribute<'a>(
    attribute: &oxc_ast::ast::JSXAttribute<'a>,
    ctx: &LintContext<'a>,
) {
    let JSXAttributeName::Identifier(attribute_name) = &attribute.name else {
        return;
    };
    if !MOTION_ANIMATE_PROPS.contains(&attribute_name.name.as_str()) {
        return;
    }
    let Some(JSXAttributeValue::ExpressionContainer(container)) = &attribute.value else {
        return;
    };
    let Some(Expression::ObjectExpression(animation)) = container.expression.as_expression()
    else {
        return;
    };
    for property in object_properties(animation) {
        let Some(property_name) = property_key_identifier_name(&property.key) else {
            continue;
        };
        if !matches!(property_name, "filter" | "backdropFilter" | "WebkitBackdropFilter") {
            continue;
        }
        let Expression::StringLiteral(value) = &property.value else {
            continue;
        };
        let Some(blur_radius) = get_blur_radius(value.value.as_str()) else {
            continue;
        };
        if blur_radius > 10.0 {
            ctx.diagnostic(
                OxcDiagnostic::warn(format!(
                    "Large animated blurs can use significant GPU memory on phones because blur({blur_radius}px) gets heavier as the blur and element grow. Use a smaller blur or a smaller element."
                ))
                .with_label(property.span),
            );
        }
    }
}

fn check_web_animation_call<'a>(
    call_expression: &oxc_ast::ast::CallExpression<'a>,
    ctx: &LintContext<'a>,
) {
    let Some(member_expression) = call_expression.callee.as_member_expression() else {
        return;
    };
    if member_expression.static_property_name() != Some("animate")
        || !is_proven_dom_event_target(member_expression.object(), ctx, &mut Vec::new())
    {
        return;
    }
    let Some(keyframes) = call_expression
        .arguments
        .first()
        .and_then(oxc_ast::ast::Argument::as_expression)
    else {
        return;
    };
    match keyframes.get_inner_expression() {
        Expression::ObjectExpression(keyframe) => check_web_animation_keyframe(keyframe, ctx),
        Expression::ArrayExpression(keyframes) => {
            for keyframe in keyframes.elements.iter().filter_map(|element| element.as_expression()) {
                if let Expression::ObjectExpression(keyframe) = keyframe.get_inner_expression() {
                    check_web_animation_keyframe(keyframe, ctx);
                }
            }
        }
        _ => {}
    }
}

fn check_web_animation_keyframe<'a>(
    keyframe: &ObjectExpression<'a>,
    ctx: &LintContext<'a>,
) {
    for property in object_properties(keyframe) {
        let property_name = if property.computed {
            match &property.key {
                PropertyKey::StringLiteral(property_name) => {
                    Some(Cow::Borrowed(property_name.value.as_str()))
                }
                _ => None,
            }
        } else {
            property.key.static_name()
        };
        let Some(property_name) = property_name else {
            continue;
        };
        if !matches!(
            property_name.as_ref(),
            "filter"
                | "backdropFilter"
                | "backdrop-filter"
                | "WebkitBackdropFilter"
                | "-webkit-backdrop-filter"
        ) {
            continue;
        }
        let Expression::StringLiteral(value) = &property.value else {
            continue;
        };
        let Some(blur_radius) = get_blur_radius(value.value.as_str()) else {
            continue;
        };
        if blur_radius > 10.0 {
            ctx.diagnostic(
                OxcDiagnostic::warn(format!(
                    "This Web Animation uses blur({blur_radius}px), which can consume significant GPU memory. Use a smaller blur or animate opacity and transform instead."
                ))
                .with_label(property.span),
            );
        }
    }
}

fn object_properties<'a>(
    object: &'a ObjectExpression<'a>,
) -> impl Iterator<Item = &'a ObjectProperty<'a>> {
    object.properties.iter().filter_map(|property| match property {
        ObjectPropertyKind::ObjectProperty(property) => Some(property.as_ref()),
        ObjectPropertyKind::SpreadProperty(_) => None,
    })
}

fn get_blur_radius(value: &str) -> Option<f64> {
    BLUR_VALUE_PATTERN
        .captures(value)
        .and_then(|captures| captures.get(1))
        .and_then(|radius| radius.as_str().parse().ok())
}
